core/src/main/scala/org/apache/spark/util/ThreadUtils.scala

```diff
@@ -17,21 +17,106 @@
 
 package org.apache.spark.util
 
+import java.util
 import java.util.concurrent._
 import java.util.concurrent.locks.ReentrantLock
 
+import com.google.common.util.concurrent.{MoreExecutors, ThreadFactoryBuilder}
 import scala.concurrent.{Awaitable, ExecutionContext, ExecutionContextExecutor, Future}
 import scala.concurrent.duration.{Duration, FiniteDuration}
 import scala.language.higherKinds
 import scala.util.control.NonFatal
 
-import com.google.common.util.concurrent.ThreadFactoryBuilder
-
 import org.apache.spark.SparkException
 import org.apache.spark.rpc.RpcAbortException
 
 private[spark] object ThreadUtils {
 
+  object MDCAwareThreadPoolExecutor {
+    def newCachedThreadPool(threadFactory: ThreadFactory): ThreadPoolExecutor = {
+      // The values needs to be synced with `Executors.newCachedThreadPool`
+      new MDCAwareThreadPoolExecutor(
+        0,
+        Integer.MAX_VALUE,
+        60L,
+        TimeUnit.SECONDS,
+        new SynchronousQueue[Runnable],
+        threadFactory)
+    }
+
+    def newFixedThreadPool(nThreads: Int, threadFactory: ThreadFactory): ThreadPoolExecutor = {
+      // The values needs to be synced with `Executors.newFixedThreadPool`
+      new MDCAwareThreadPoolExecutor(
+        nThreads,
+        nThreads,
+        0L,
+        TimeUnit.MILLISECONDS,
+        new LinkedBlockingQueue[Runnable],
+        threadFactory)
+    }
+
+    def newSingleThreadExecutor(threadFactory: ThreadFactory): ExecutorService = {
+      // The values needs to be synced with `Executors.newSingleThreadExecutor`
+      Executors.unconfigurableExecutorService(
+        new MDCAwareThreadPoolExecutor(
```

Comment:
I'm not sure, is there any reason here not to use previous`FinalizableDelegatedExecutorService`?

Reply:
The only difference now is the finalize method. So I don't see any benefit from it.
In order to propagate MDC to all places, we need to pass it to all threads.

Reply:
I'd keep original FinalizableDelegatedExecutorService, if not special reason to change it. Cannot we wrap MDCAwareThreadPoolExecutor by a FinalizableDelegatedExecutorService here too?

Reply:
FinalizableDelegatedExecutorService is package protected so we can't create it. I can copy the code if you insist.

Reply:
The finalize method helps to shutdown the underlying thread pool. I'm not sure if we rely on this or not in Spark. If you are confident on this change, then it should be fine.

Reply:
But let's at least leave a comment here in case people use it without knowing there's difference vs built-in `newSingleThreadExecutor`.

Reply:
@viirya @HyukjinKwon 
1. Yes I am confident: finalize is "best-effort" the JVM does guaranty to run finalize also in later JVM it becomes 
```
@Deprecated(since="9")
    protected void finalize() throws Throwable { }
```
2. OK I will add comment even so I think in any case no one should rely on finalize.